allow specifying dependencies packages

galexrt opened the pull request:
This allows to install on CentOS, Fedora and other OSes by specifying
the packages manually.
Example:
Fedora packages would be `crond`, `git` and `wget`.

Signed-off-by: Alexander Trost <galexrt@googlemail.com>

***

One potential issue might be the switch to use `package:` module instead of `apt` though that could be altered to use `apt:` module when the package manager is `apt` on the system. For other OSes where the cache(s) are automatically refreshed the `package:` could be used.
What do you think?

nickjj commented:
Hi,

I really do appreciate the PR but since I don't use centos, fedora and others I'm a bit reluctant to make the role more complex to configure for the supported distros (debian and ubuntu).

galexrt commented:
@nickjj the current default value list of packages is for Debian / Ubuntu.

The `acme_sh_dependencies` var only needs to be changed from its default when the package names are different (e.g., when using CentOS / Fedora the cron package is named `crond`).

Yes, it adds more complexity (ignoring the second task for non apt based OSes using `package:` module, I have added to show it off now) but in the end it allows other OSes to work fine with this role.

In my case Fedora only needs a different package for cron because the package is named `crond (`acme_sh_dependencies: ["crond", "git", "wget"]`).

***

As mentioned in the brackets text, for showing the newly added task for non-apt OSes, I have added the conditional for apt based and other pkg managers to the PR now.

jhomer-hscl commented:
> @nickjj Thanks for your comment! The current default value list of packages is for Debian / Ubuntu.
> 
> The `acme_sh_dependencies` var only needs to be changed from its default when the package names are different (e.g., when using CentOS / Fedora the cron package is named `crond`).
> 
> Yes, it adds more complexity (ignoring the second task for non apt based OSes using `package:` module, I have added to show it off now) but in the end it allows other OSes to work fine with this role.
> 
> In my case Fedora only needs a different package for cron because the package is named `crond (`acme_sh_dependencies: ["crond", "git", "wget"]`).
> 
> As mentioned in the brackets text, for showing the newly added task for non-apt OSes, I have added the conditional for apt based and other pkg managers to the PR now.

Was looking at adding similar functionality but spotted it has already been done. Configurable list of packages is good idea as they appear to be slightly different for CentOS vs Fedora. Happy to confirm tests on CentOS 7/8 variants if this convinces @nickjj to accept the PR.

nickjj commented:
I don't have much experience with the `package` module since I typically use `apt`. Do other distros not have the concept of a cache time?

jhomer-hscl commented:
Both yum and dnf have this concept. Looks like the defaults are 6 and 48 hours respectively.

The associated ansible modules support `update_cache` that can force an update of the local cache. But neither appear to support `cache_valid_time` like the apt module does.

An alternative could be to use `ansible_pkg_mgr` as described here: [Install Packages](https://ansible-tips-and-tricks.readthedocs.io/en/latest/os-dependent-tasks/installing_packages/) to call the correct local package management tool. Would need to drop the `cache_valid_time` as not supported by yum/dnf. I have not checked to see if this works yet.

galexrt commented:
@jhomer-hscl Or as the PR is right now, for systems using `apt` as the package manager the `apt` module is used and for other package managers the `package` module is used.

nickjj commented:
Right, I brought it up because I was wondering if we could come up with a solution that doesn't require having 2 tasks to install the packages depending on what distro you have.

jhomer-hscl commented:
Of course. If there is no need to force an update of the cache (or another function provided by `apt'. `yum`, etc. modules but that is not present in `package`) then it looks like `package` could cover apt and yum based systems and not complicate things.

@nickjj is there a reason to force a cache update in this scenario. If there was a need to add a new repository then I can see why this would be desirable.

nickjj commented:
Back when this role was created I was a fan of updating caches at the role level but for the last few years not so much. Just bootstrapping a new server to install Python for Ansible would require running an apt-get update initially.

If you want to remove the apt task and all traces of the cache variable I'd merge it in.

galexrt commented:
@nickjj @jhomer-hscl I have updated the PR to use the `package` module instead of `apt` now and removed the cache variable.

The CI just failed. That seems to be caused by the apt cache not being "updated".
Should I update the test tasks to include an apt cache update or how should I proceed?

nickjj commented:
Yeah it would be common to add the update to the tests file. It's what I do in all of my more recent roles.

You can drop this into the pre-tasks in the `tests.yml` file:

```yml
    - name: Run the equivalent of "apt-get update"
      apt:
        update_cache: true
      changed_when: false
```

galexrt commented:
Here we go, CI is passing now! :slightly_smiling_face:

nickjj commented:
Thanks. Would you mind resubmitting this with these 2 changes:

1. Adjusting the commit message to be a single line of `Allow specifying a list of dependency packages`

2. Changing the README comment to be:

```md
FROM:
# acme.sh package depedencies to install. The packages listed here are for
# Debian based systems (Debian and Ubuntu). You need to change this for other
# OSes like, Fedora, CentOS and so on.

TO:
# acme.sh package dependencies. The default values are for Debian / Ubuntu.
# For CentOS and Fedora you can replace "cron" with "crond".
acme_sh_dependencies: ["cron", "git", "wget"]
```

galexrt commented:
@nickjj I have updated the README variable text and also adjusted the commit message to be a single line (besides the sign-off, I can remove it as well if wanted).

nickjj commented:
Thanks!
